internal/proxymanager/proxymanager_test.go

```diff
@@ -0,0 +1,160 @@
+		name := name
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			_, err := proxymanager.New(context.Background(), tc.config)
+			if tc.wantErr {
+				// We don't return here as we want to check that the apparmor
+				// dir is in the expected state even in error cases
+				require.Error(t, err, "New should have failed but didn't")
+				return
+			}
+			require.NoError(t, err, "New failed but shouldn't have")
+		})
+	}
+}
+
+func TestApply(t *testing.T) {
+	t.Parallel()
+
+	envConfigPath := filepath.Join("environment.d", "99adsys-proxy.conf")
+	initialTime := time.Date(2001, 1, 1, 1, 1, 1, 1, time.UTC)
+
+	tests := map[string]struct {
+		config proxymanager.Config
+
+		existingDirs       []string
+		prevContents       map[string]string
+		readOnlyFiles      []string
+		wantUnchangedFiles []string
+
+		wantErr bool
+	}{
+		"No options set":   {config: proxymanager.Config{}},
+		"HTTP option set":  {config: proxymanager.Config{HTTPProxy: "http://example.com:8080"}},
+		"Some options set": {config: proxymanager.Config{HTTPProxy: "http://example.com:8080", HTTPSProxy: "https://example.com:8080"}},
+		"Some options set, environment parent directories are created": {
+			config:       proxymanager.Config{HTTPProxy: "http://example.com:8080", HTTPSProxy: "https://example.com:8080"},
+			existingDirs: []string{},
+		},
+		"Some options set, environment file is unchanged": {
+			wantUnchangedFiles: []string{envConfigPath},
+			config:             proxymanager.Config{HTTPProxy: "http://example.com:8080", HTTPSProxy: "https://example.com:8080"},
+			prevContents: map[string]string{envConfigPath: `### This file was generated by ADSys - manual changes will be overwritten
+HTTP_PROXY=http://example.com:8080
+http_proxy=http://example.com:8080
+HTTPS_PROXY=https://example.com:8080
+https_proxy=https://example.com:8080
+`},
+		},
+		"All options set": {config: proxymanager.Config{
+			HTTPProxy: "http://example.com:8080", HTTPSProxy: "https://example.com:8080",
+			FTPProxy: "ftp://example.com:8080", SocksProxy: "socks://example.com:8080",
+			NoProxy: "localhost,127.0.0.1",
+		}},
+		"Password is escaped":                          {config: proxymanager.Config{HTTPProxy: "http://username:p@$$:w0rd@example.com:8080"}},
+		"Escaped password is not escaped again":        {config: proxymanager.Config{HTTPProxy: "http://username:p%40$$%3Aw0rd@example.com:8080"}},
+		"Domain username is escaped":                   {config: proxymanager.Config{HTTPProxy: `http://EXAMPLE\bobsmith:p@$$:w0rd@example.com:8080`}},
+		"Domain username without password is escaped":  {config: proxymanager.Config{HTTPProxy: `http://EXAMPLE\bobsmith@example.com:8080`}},
+		"Escaped domain username is not escaped again": {config: proxymanager.Config{HTTPProxy: `http://EXAMPLE%5Cbobsmith@example.com:8080`}},
+
+		// Error cases
+		"Environment config file is read-only": {readOnlyFiles: []string{envConfigPath}, prevContents: map[string]string{envConfigPath: "something"}, wantErr: true},
```

Comment:
Don't forget that our error cases are supposed to be prefixed with `Error when ...`, `Error on ...`, etc.